### DISCUSSION TITLE
Stop evaluating macros

### DISCUSSION THREAD

**Discussion by SuperFola** [ArkScript-lang/Ark#473] [Workgroup]
In the current implementation of the language and macro processor, every argument passed to a macro function is evaluated as a macro if possible.

However this is a problem for some nodes that ressemble macros but aren't, eg `(@ list index)` the index could be a node index in the list instead of an element index.

There are multiple solutions to this:
- add quoting, quasiquoting and unquoting (hardest solution) like Lisp and Scheme do it
- add a way to stop argument evaluation on demand (easy)

I went with the second solution, see: https://github.com/ArkScript-lang/Ark/issues/472 and would like to know what you think about it

**Comment by metaleap** [comment]
> quasiquoting and unquoting (hardest solution)

Not hard, steal [mine](https://github.com/metaleap/go-lisp/blob/main/special_forms.go#L203) which follows [this](https://github.com/kanaka/mal/issues/643#issue-1838234677)  =)

To be able to quasiQuote/unquote is the coolest for code (and other lists) constructions, can't even imagine having a macro facility on the one hand but not those on the other?..

**Reply by SuperFola** [reply]
> To be able to quasiQuote/unquote is the coolest for code (and other lists) constructions

It is! For now I went with a construction that stops evaluation (badly named `$paste`, can't call it quoting as it would quote `($paste a)` like lisp does but rather generate a symbol a that would be evaluated by the compiler), which achieve nearly the same?

I'll have a look, thanks for the link!

**Reply by metaleap** [reply]
> I'll have a look, thanks for the link!

Since that's based on MAL it has 2 list types (list, vec) so don't let the "vec" stuff confuse ya there.

This is a "special form" aka lazy-eval'd-builtin so it can call eval/apply right in there, making for a trivial impl that works trivially for a tree-walking interpreter but is probably no solution for writing a Lisp compiler — whether to native code or custom byte-code such as yours. Unless you anyway do eval/apply at compile time for macro-running, then it might be applicable. If it isn't, then maybe that's what you meant by "hardest solution" and sorry for my initial cockiness :grin:
